File: tools/fixtures/toml-to-pkl.mjs
#!/usr/bin/env node
// Convert a Vize TOML test fixture into an equivalent Apple Pkl module, matching
// the hand-authored convention in tests/fixtures/sfc/basic.pkl:
//
//   - leading `#` comment lines become `//` comments
//   - top-level scalars (mode/test_type) stay as plain assignments
//   - arrays of tables become `new Listing { new { ... } }`
//   - multi-line strings use Pkl `"""` blocks whose content is indented to the
//     key's column so Pkl's indentation stripping restores the original text
//     verbatim; a value ending in a newline gets a trailing blank line so the
//     newline round-trips exactly.
//
// The result is byte-exact: the Rust snapshot / expected-output tests pass
// without regenerating snapshots, which is what verifies the conversion.
//
// Usage: node tools/fixtures/toml-to-pkl.mjs <fixture.toml> [more.toml ...]

import { readFileSync, writeFileSync } from "node:fs";
import TOML from "@iarna/toml";

const TQ = '"'.repeat(3);

function leadingComment(text) {
  const out = [];
  for (const line of text.split("\n")) {
    const s = line.trim();
    if (s.startsWith("#")) out.push("//" + s.slice(1));
    else if (s === "") {
      if (out.length) break;
    } else break;
  }
  return out;
}

function pklInlineString(value) {
  // A plain string is fine when there is nothing to escape and no `\(...)`
  // interpolation to worry about; otherwise fall back to a raw `#"..."#` fence.
  if (!/["\\]/.test(value)) return `"${value}"`;
  let n = 1;
  while (value.includes('"' + "#".repeat(n))) n++;
  const fence = "#".repeat(n);
  return `${fence}"${value}"${fence}`;
}

function pklMultiline(value, indent) {
  const pad = " ".repeat(indent);
  const lines = value.split("\n");
  const body = lines.length > 1 && lines[lines.length - 1] === "" ? lines.slice(0, -1) : lines;
  const out = [TQ];
  for (const ln of body) out.push(ln ? pad + ln : "");
  if (value.endsWith("\n")) out.push("");
  out.push(pad + TQ);
  return out;
}

function emitValue(value, indent, out) {
  // out[-1] already ends with `<key> = ` (or `new `); append the value.
  const pad = "  ".repeat(indent);
  const last = out.length - 1;
  if (typeof value === "boolean") {
    out[last] += value ? "true" : "false";
  } else if (typeof value === "number" || typeof value === "bigint") {
    out[last] += String(value);
  } else if (typeof value === "string") {
    if (value.includes("\n")) {
      const keyCol = out[last].length - out[last].replace(/^ +/, "").length;
      const rendered = pklMultiline(value, keyCol);
      out[last] += rendered[0];
      out.push(...rendered.slice(1));
    } else {
      out[last] += pklInlineString(value);
    }
  } else if (Array.isArray(value)) {
    out[last] += "new Listing {";
    for (const child of value) {
      if (child && typeof child === "object" && !Array.isArray(child)) {
        out.push(`${pad}  new {`);
        for (const [k, v] of Object.entries(child)) {
          out.push(`${pad}    ${k} = `);
          emitValue(v, indent + 2, out);
        }
        out.push(`${pad}  }`);
      } else {
        out.push(`${pad}  `);
        emitValue(child, indent + 1, out);
      }
    }
    out.push(`${pad}}`);
  } else if (value && typeof value === "object") {
    out[last] += "new {";
    for (const [k, v] of Object.entries(value)) {
      out.push(`${pad}  ${k} = `);
      emitValue(v, indent + 1, out);
    }
    out.push(`${pad}}`);
  } else {
    throw new Error(`unsupported value: ${JSON.stringify(value)}`);
  }
}

function convert(tomlPath, amends) {
  const text = readFileSync(tomlPath, "utf8");
  const data = TOML.parse(text);
  const out = leadingComment(text);
  if (out.length) out.push("");
  if (amends) {
    // `amends` makes the fixture a typed instance of the schema module: Pkl
    // checks every assigned property/field against it.
    out.push(`amends ${JSON.stringify(amends)}`);
    out.push("");
  }
  for (const [key, value] of Object.entries(data)) {
    if (amends && Array.isArray(value)) {
      // Amend the schema-declared Listing so its `new {}` elements are typed
      // against the element class (e.g. `Case`) rather than `Dynamic`.
      out.push(`${key} {`);
      for (const child of value) {
        if (child && typeof child === "object" && !Array.isArray(child)) {
          out.push("  new {");
          for (const [k, v] of Object.entries(child)) {
            out.push(`    ${k} = `);
            emitValue(v, 2, out);
          }
          out.push("  }");
        } else {
          out.push("  ");
          emitValue(child, 1, out);
        }
      }
      out.push("}");
    } else {
      out.push(`${key} = `);
      emitValue(value, 0, out);
    }
    out.push("");
  }
  while (out.length && out[out.length - 1] === "") out.pop();
  return out.join("\n") + "\n";
}

const rawArgs = process.argv.slice(2);
let amends = null;
const args = [];
for (const a of rawArgs) {
  if (a.startsWith("--amends=")) amends = a.slice("--amends=".length);
  else args.push(a);
}
if (args.length === 0) {
  console.error("usage: node toml-to-pkl.mjs [--amends=Schema.pkl] <fixture.toml> ...");
  process.exit(2);
}
for (const arg of args) {
  const pklPath = arg.replace(/\.toml$/, ".pkl");
  writeFileSync(pklPath, convert(arg, amends));
  console.log(`${arg} -> ${pklPath}`);
}

File: tools/commands/fixtures/toml-to-pkl.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [dependencies]
//! serde = { version = "1", features = ["derive"] }
//! serde_json = "1"
//! toml_edit = "0.22"
//!
//! [package]
//! edition = "2024"
//! ```

#[path = "../../rust/common.rs"]
mod common;

use std::{env, path::Path, process::ExitCode};
use toml_edit::{Array, DocumentMut, Item, Table, Value};

fn main() -> ExitCode {
    common::main_result(run())
}

fn run() -> Result<(), String> {
    let mut amends = None;
    let mut files = Vec::new();
    for arg in env::args().skip(1) {
        if let Some(value) = arg.strip_prefix("--amends=") {
            amends = Some(value.to_string());
        } else {
            files.push(arg);
        }
    }
    if files.is_empty() {
        return Err("usage: rust-script tools/commands/fixtures/toml-to-pkl.rs [--amends=Schema.pkl] <fixture.toml> ...".to_string());
    }
    for file in files {
        let pkl = convert(Path::new(&file), amends.as_deref())?;
        let pkl_path = file.strip_suffix(".toml").unwrap_or(&file).to_string() + ".pkl";
        common::write_text(&pkl_path, &pkl)?;
        println!("{file} -> {pkl_path}");
    }
    Ok(())
}

fn convert(toml_path: &Path, amends: Option<&str>) -> Result<String, String> {
    let text = common::read_text(toml_path)?;
    let data = text
        .parse::<DocumentMut>()
        .map_err(|error| format!("cannot parse {}: {error}", toml_path.display()))?;
    let mut out = leading_comment(&text);
    if !out.is_empty() {
        out.push(String::new());
    }
    if let Some(amends) = amends {
        out.push(format!("amends {}", pkl_inline_string(amends)));
        out.push(String::new());
    }
    for (key, value) in data.as_table().iter() {
        if amends.is_some() && value.as_array_of_tables().is_some() {
            out.push(format!("{key} {{"));
            let array = value.as_array_of_tables().unwrap();
            for child in array.iter() {
                out.push("  new {".to_string());
                for (child_key, child_value) in child.iter() {
                    out.push(format!("    {child_key} = "));
                    emit_item(child_value, 2, &mut out)?;
                }
                out.push("  }".to_string());
            }
            out.push("}".to_string());
        } else {
            out.push(format!("{key} = "));
            emit_item(value, 0, &mut out)?;
        }
        out.push(String::new());
    }
    while out.last().map(|line| line.is_empty()).unwrap_or(false) {
        out.pop();
    }
    Ok(format!("{}\n", out.join("\n")))
}

fn leading_comment(text: &str) -> Vec<String> {
    let mut out = Vec::new();
    for line in text.split('\n') {
        let trimmed = line.trim();
        if let Some(comment) = trimmed.strip_prefix('#') {
            out.push(format!("//{comment}"));
        } else if trimmed.is_empty() {
            if !out.is_empty() {
                break;
            }
        } else {
            break;
        }
    }
    out
}

fn pkl_inline_string(value: &str) -> String {
    if !value.contains('"') && !value.contains('\\') {
        return format!("\"{value}\"");
    }
    let mut fence_len = 1usize;
    while value.contains(&format!("\"{}", "#".repeat(fence_len))) {
        fence_len += 1;
    }
    let fence = "#".repeat(fence_len);
    format!("{fence}\"{value}\"{fence}")
}

fn pkl_multiline(value: &str, indent: usize) -> Vec<String> {
    let pad = " ".repeat(indent);
    let lines = value.split('\n').collect::<Vec<_>>();
    let body = if lines.len() > 1 && lines.last() == Some(&"") {
        &lines[..lines.len() - 1]
    } else {
        &lines[..]
    };
    let mut out = vec!["\"\"\"".to_string()];
    for line in body {
        if line.is_empty() {
            out.push(String::new());
        } else {
            out.push(format!("{pad}{line}"));
        }
    }
    if value.ends_with('\n') {
        out.push(String::new());
    }
    out.push(format!("{pad}\"\"\""));
    out
}

fn emit_item(item: &Item, indent: usize, out: &mut Vec<String>) -> Result<(), String> {
    if let Some(value) = item.as_value() {
        return emit_value(value, indent, out);
    }
    if let Some(array) = item.as_array_of_tables() {
        return emit_array_of_tables(array.iter(), indent, out);
    }
    if let Some(table) = item.as_table() {
        return emit_table(table, indent, out);
    }
    Err(format!("unsupported TOML item: {item:?}"))
}

fn emit_value(value: &Value, indent: usize, out: &mut Vec<String>) -> Result<(), String> {
    let pad = "  ".repeat(indent);
    let last = out.len() - 1;
    if let Some(value) = value.as_bool() {
        out[last].push_str(if value { "true" } else { "false" });
    } else if let Some(value) = value.as_integer() {
        out[last].push_str(&value.to_string());
    } else if let Some(value) = value.as_float() {
        out[last].push_str(&value.to_string());
    } else if let Some(value) = value.as_str() {
        if value.contains('\n') {
            let key_col = out[last].len() - out[last].trim_start_matches(' ').len();
            let rendered = pkl_multiline(value, key_col);
            out[last].push_str(&rendered[0]);
            out.extend(rendered.into_iter().skip(1));
        } else {
            out[last].push_str(&pkl_inline_string(value));
        }
    } else if let Some(array) = value.as_array() {
        emit_array(array, indent, out)?;
    } else {
        return Err(format!("unsupported TOML value: {value:?}"));
    }
    if matches!(value, Value::Array(_)) {
        let _ = pad;
    }
    Ok(())
}

fn emit_array(array: &Array, indent: usize, out: &mut Vec<String>) -> Result<(), String> {
    let pad = "  ".repeat(indent);
    let last = out.len() - 1;
    out[last].push_str("new Listing {");
    for child in array.iter() {
        out.push(format!("{pad}  "));
        emit_value(child, indent + 1, out)?;
    }
    out.push(format!("{pad}}}"));
    Ok(())
}

fn emit_array_of_tables<'a>(
    array: impl Iterator<Item = &'a Table>,
    indent: usize,
    out: &mut Vec<String>,
) -> Result<(), String> {
    let pad = "  ".repeat(indent);
    let last = out.len() - 1;
    out[last].push_str("new Listing {");
    for child in array {
        out.push(format!("{pad}  new {{"));
        for (key, value) in child.iter() {
            out.push(format!("{pad}    {key} = "));
            emit_item(value, indent + 2, out)?;
        }
        out.push(format!("{pad}  }}"));
    }
    out.push(format!("{pad}}}"));
    Ok(())
}

fn emit_table(table: &Table, indent: usize, out: &mut Vec<String>) -> Result<(), String> {
    let pad = "  ".repeat(indent);
    let last = out.len() - 1;
    out[last].push_str("new {");
    for (key, value) in table.iter() {
        out.push(format!("{pad}  {key} = "));
        emit_item(value, indent + 1, out)?;
    }
    out.push(format!("{pad}}}"));
    Ok(())
}
